File: admin_desktop/lib/core/entities/producto.dart
class RecetaItem {
  final String insumoId;
  final double cantidad;

  RecetaItem({required this.insumoId, required this.cantidad});

  factory RecetaItem.fromJson(Map<String, dynamic> json) {
    return RecetaItem(
      insumoId: json['insumoId'],
      cantidad: (json['cantidad'] as num).toDouble(),
    );
  }

  Map<String, dynamic> toJson() => {'insumoId': insumoId, 'cantidad': cantidad};
}

class Producto {
  final String id;
  final String nombre;
  final String descripcion;
  final String categoria;
  final double precio;
  final List<RecetaItem> receta;
  final bool activo;

  Producto({
    required this.id,
    required this.nombre,
    required this.descripcion,
    required this.categoria,
    required this.precio,
    required this.receta,
    this.activo = true,
  });

  factory Producto.fromJson(Map<String, dynamic> json) {
    return Producto(
      id: json['id'] ?? '',
      nombre: json['nombre'],
      descripcion: json['descripcion'],
      categoria: json['categoria'] ?? '',
      precio: (json['precio'] as num).toDouble(),
      receta: (json['receta'] as List)
          .map((e) => RecetaItem.fromJson(e))
          .toList(),
      activo: json['activo'] ?? true,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'nombre': nombre,
      'descripcion': descripcion,
      'categoria': categoria,
      'precio': precio,
      'receta': receta.map((e) => e.toJson()).toList(),
      'activo': activo,
    };
  }
}

File: functions/src/domain/entities/producto.ts
export interface RecetaItem {
    insumoId: string;
    cantidad: number; // Cantidad requerida por insumo
}

export class Producto {
    constructor(
        public readonly id: string,
        public nombre: string,
        public descripcion: string,
        public precio: number,
        public receta: RecetaItem[],
        public activo: boolean = true
    ) { }

    static create(props: {
        id: string;
        nombre: string;
        descripcion: string;
        precio: number;
        receta: RecetaItem[];
    }): Producto {
        if (props.precio < 0) throw new Error("Price cannot be negative");
        return new Producto(
            props.id,
            props.nombre,
            props.descripcion,
            props.precio,
            props.receta
        );
    }

    // Validar si hay stock suficiente dado un map de stock actual
    validarStock(stockMap: Map<string, number>): boolean {
        for (const item of this.receta) {
            const stock = stockMap.get(item.insumoId) || 0;
            if (stock < item.cantidad) return false;
        }
        return true;
    }
}
